lbrynet/reflector/server/server.py

```diff
@@ -11,27 +11,89 @@
 
 log = logging.getLogger(__name__)
 
+SEND_SD_BLOB = 'send_sd_blob'
+SEND_BLOB = 'send_blob'
+RECEIVED_SD_BLOB = 'received_sd_blob'
+RECEIVED_BLOB = 'received_blob'
+NEEDED_BLOBS = 'needed_blobs'
+VERSION = 'version'
+BLOB_SIZE = 'blob_size'
+BLOB_HASH = 'blob_hash'
+SD_BLOB_SIZE = 'sd_blob_size'
+SD_BLOB_HASH = 'sd_blob_hash'
+
 
 class ReflectorServer(Protocol):
     def connectionMade(self):
         peer_info = self.transport.getPeer()
         log.debug('Connection made to %s', peer_info)
         self.peer = self.factory.peer_manager.get_peer(peer_info.host, peer_info.port)
         self.blob_manager = self.factory.blob_manager
+        self.protocol_version = self.factory.protocol_version
         self.received_handshake = False
-        self.received_descriptor = False
-        self.protocol_version = REFLECTOR_V2
         self.peer_version = None
         self.receiving_blob = False
         self.incoming_blob = None
         self.blob_write = None
-        self.blob_finished_d = None
+        self.blob_f_d = None
         self.cancel_write = None
         self.request_buff = ""
 
     def connectionLost(self, reason=failure.Failure(error.ConnectionDone())):
         log.info("Reflector upload from %s finished" % self.peer.host)
 
+    def handle_error(self, err):
+        log.error(err.getTraceback())
+        self.transport.loseConnection()
+
+    def send_response(self, response_dict):
+        self.transport.write(json.dumps(response_dict))
+
+    ############################
+    # Incoming blob file stuff #
+    ############################
+
+    def clean_up_failed_upload(self, err, blob):
+        log.warning("Failed to receive %s", blob.blob_hash[:16])
+        if err.check(DownloadCanceledError):
+            err.trap(DownloadCanceledError)
+            self.blob_manager.delete_blobs([blob.blob_hash])
+        else:
+            log.exception(err)
+        return False
+
+    def log_received_blob(self, blob):
+        log.info("Received %s", blob.blob_hash[:16])
+        return True
+
+    def handle_incoming_blob(self, response_key):
+        """
+        Open blob for writing and send a response indicating if the transfer was
+        successful when finished.
+
+        response_key will either be received_blob or received_sd_blob
+        """
+
+        blob = self.incoming_blob
+        self.blob_f_d, self.blob_write, self.cancel_write = blob.open_for_writing(self.peer)
+        self.blob_f_d.addCallback(lambda _: self.blob_manager.blob_completed(blob))
+        self.blob_f_d.addCallback(lambda _: self.close_blob())
+        self.blob_f_d.addCallbacks(lambda _: self.log_received_blob(blob),
+                                   lambda err: self.clean_up_failed_upload(err, blob))
+        self.blob_f_d.addCallback(lambda result: {response_key: result})
+        self.blob_f_d.addCallback(self.send_response)
```

Comment:
I think you changed blob_finished_d to blob_f_d because blob_finished_d made this section too long. Instead, try:

```
blob = self.incoming_blob
self.blob_f_d, self.blob_write, self.cancel_write = blob.open_for_writing(self.peer)
self.blob_finished_d.addCallback(self._on_completed_blob, response_key)
```

```
@defer.inlineCallbacks
def _on_completed_blob(self, response_key):
    blob = self.incoming_blob
    try:
        yield self.blob_manager.blob_completed(blob)
        yield self.close_blob()
        log.info("Received %s", blob)
        yield self.send_respond({response_key: True})
    except Exception as e:
        yield self.clean_up_failed_upload(e, blob)
        yield self.send_respond({response_key: False})
```